### PR TITLE
MASTERDIET-21 - Creating BMI calculator

### PR DESCRIPTION
**Pull request by jakubowiczish**

### CODE REVIEW

File: src/main/java/pl/agh/edu/master_diet/controller/CalculatorController.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+@RestController
+@RequestMapping("/calculator")
+@AllArgsConstructor(onConstructor = @__(@Autowired))
+public class CalculatorController {
+
+    private final CalculatorService calculatorService;
+    private final ConversionService conversionService;
+
+    @PostMapping("/bmi")
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/controller/CalculatorController.java:26]: it's shortcut from Body Mass index should be BMI

File: src/main/java/pl/agh/edu/master_diet/controller/CalculatorController.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    public ResponseEntity<CalculateBmiResponse> calculateBmi(@Valid @RequestBody final CalculateBmiRequest request) {
+        final BmiParameters bmiParameters = conversionService.convert(request);
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/controller/CalculatorController.java:28]: calculateBMI, BMIParameters

File: src/main/java/pl/agh/edu/master_diet/core/model/rest/calculator/bmi/CalculateBmiRequest.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class CalculateBmiRequest {
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/core/model/rest/calculator/bmi/CalculateBmiRequest.java:12]: CalculateBMIRequest

File: src/main/java/pl/agh/edu/master_diet/core/model/rest/calculator/bmi/CalculateBmiResponse.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import pl.agh.edu.master_diet.core.model.shared.WeightRange;
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class CalculateBmiResponse {
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/core/model/rest/calculator/bmi/CalculateBmiResponse.java:13]: CalculateBMIResponse

File: src/main/java/pl/agh/edu/master_diet/core/model/rest/calculator/bmi/CalculateBmiResponse.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    private Double bmi;
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/core/model/rest/calculator/bmi/CalculateBmiResponse.java:15]: userBMI

File: src/main/java/pl/agh/edu/master_diet/core/model/shared/BmiParameters.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class BmiParameters {
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/core/model/shared/BmiParameters.java:12]: BMIParametes

File: src/main/java/pl/agh/edu/master_diet/core/model/shared/WeightRange.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+public enum WeightRange {
+
+    UNDERWEIGHT(18.49d),
+    NORMAL(24.9d),
+    OVERWEIGHT(29.9d),
+    OBESE(Float.MAX_VALUE);
+
+    private final double upperLimit;
+
+    public static WeightRange getWeightRangeFromBmi(final double bmi) {
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/core/model/shared/WeightRange.java:18]: getWeightRangeFromBMI

Comment (RafisSomeone) [reply src/main/java/pl/agh/edu/master_diet/core/model/shared/WeightRange.java:18]: final double userBMI

File: src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import org.apache.commons.math3.util.Precision;
+import org.springframework.stereotype.Service;
+import pl.agh.edu.master_diet.core.model.rest.calculator.bmi.CalculateBmiResponse;
+import pl.agh.edu.master_diet.core.model.shared.BmiParameters;
+import pl.agh.edu.master_diet.core.model.shared.WeightRange;
+
+@Service
+public class CalculatorService {
+
+    public CalculateBmiResponse calculateBmi(final BmiParameters bmiParameters) {
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java:12]: calculateBMI and userBMIParameters

File: src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        final Double weight = bmiParameters.getWeight();
+        final Double height = bmiParameters.getHeight() / 100.0d;
+        final double bmi = weight / (height * height);
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java:15]: userBMI

File: src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import pl.agh.edu.master_diet.core.model.rest.calculator.bmi.CalculateBmiResponse;
+import pl.agh.edu.master_diet.core.model.shared.BmiParameters;
+import pl.agh.edu.master_diet.core.model.shared.WeightRange;
+
+@Service
+public class CalculatorService {
+
+    public CalculateBmiResponse calculateBmi(final BmiParameters bmiParameters) {
+        final Double weight = bmiParameters.getWeight();
+        final Double height = bmiParameters.getHeight() / 100.0d;
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java:14]: magic number

Comment (RafisSomeone) [reply src/main/java/pl/agh/edu/master_diet/service/CalculatorService.java:14]: height / 100 it's height ? maybe change name height

File: src/main/java/pl/agh/edu/master_diet/service/converter/BmiConverter.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
+package pl.agh.edu.master_diet.service.converter;
+
+import org.springframework.stereotype.Component;
+import pl.agh.edu.master_diet.core.model.rest.calculator.bmi.CalculateBmiRequest;
+import pl.agh.edu.master_diet.core.model.shared.BmiParameters;
+
+@Component
+public class BmiConverter
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/service/converter/BmiConverter.java:8]: BMIConverter

File: src/main/java/pl/agh/edu/master_diet/service/converter/ConversionServiceImpl.java [hunk 12]
```diff
@@ -14,6 +16,7 @@
 
     private final UserParametersConverter userParametersConverter;
     private final UserPlanConverter userPlanConverter;
+    private final BmiConverter bmiConverter;
```
Comment (RafisSomeone) [inline src/main/java/pl/agh/edu/master_diet/service/converter/ConversionServiceImpl.java:19]: userBMIConverter

File: src/main/java/pl/agh/edu/master_diet/controller/CalculatorController.java (hunk 1 above)
Comment (jakubowiczish) [reply src/main/java/pl/agh/edu/master_diet/controller/CalculatorController.java]: Fixed

**Pull request merged**